drivers/led/led_gpio.c

```diff
@@ -0,0 +1,168 @@
+};
+
+struct led_dts_cfg {
+	struct gpio_cfg gpio;
+	const char *label;
+	uint32_t id;
+};
+
+struct led_gpio_data {
+	struct device *gpio_dev;
+	const struct led_dts_cfg *cfg;
+	bool state;
+};
+
+struct led_gpio_cfg {
+	const struct led_dts_cfg *led;
+	uint32_t led_cnt;
+};
+
+static inline struct led_gpio_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct led_gpio_cfg *get_dev_config(struct device *dev)
+{
+	return dev->config_info;
+}
+
+static struct led_gpio_data *led_parameters_get(struct device *dev,
+						uint32_t led)
+{
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		if (data[i].cfg->id == led) {
+			return &data[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int gpio_led_set_no_delay(struct led_gpio_data *data, bool state)
+{
+	int err;
+
+	data->state = state;
+
+	err = gpio_pin_set(data->gpio_dev, data->cfg->gpio.pin,
+			   state);
+	if (err) {
+		LOG_DBG("%s turn %s error %d",
+			data->cfg->label,
+			state ? "on" : "off",
+			err);
+	} else {
+		LOG_DBG("%s turn %s", data->cfg->label,
+			state ? "on" : "off");
+	}
+
+	return err;
+}
+
+static int gpio_led_set(struct led_gpio_data *data, bool state)
+{
+	return gpio_led_set_no_delay(data, state);
+}
+
+static int gpio_led_on(struct device *dev, uint32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, true);
+}
+
+static int gpio_led_off(struct device *dev, uint32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, false);
+}
+
+static const struct led_driver_api led_api = {
+	.on = gpio_led_on,
+	.off = gpio_led_off,
+};
+
+static int gpio_led_init(struct device *dev)
+{
+	int err;
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		data[i].gpio_dev = device_get_binding(cfg->led[i].gpio.port);
+		if (!data[i].gpio_dev) {
+			LOG_DBG("Failed to get GPIO device for port %s pin %d",
+				cfg->led[i].gpio.port, cfg->led[i].gpio.pin);
+			return -EINVAL;
+		}
+
+		/* Set all leds pin as output */
+		err = gpio_pin_configure(data[i].gpio_dev,
+					 cfg->led[i].gpio.pin,
+					 (GPIO_OUTPUT_INACTIVE |
+					  cfg->led[i].gpio.flags));
+		if (err) {
+			LOG_DBG("Failed to set GPIO port %s pin %d as output",
+				cfg->led[i].gpio.port, cfg->led[i].gpio.pin);
+			return err;
+		}
+
+
+		data[i].cfg = &cfg->led[i];
+	}
+
+	return 0;
+}
+
+static const struct led_dts_cfg led_dts[] = {DT_INST_FOREACH_CHILD(0,
+								   LED_DATA)};
+static struct led_gpio_data led_data[ARRAY_SIZE(led_dts)];
+static const struct led_gpio_cfg led_cfg = {
+	.led_cnt = ARRAY_SIZE(led_dts),
+	.led = led_dts
+};
+
+DEVICE_AND_API_INIT(gpio_led, DT_PROP(DT_DRV_INST(0), driver),
```

Comment:
I don't mind many instances of this driver. I just wonder if it's really needed. I can add support for it if you want to. Is it possible to have many `leds` node instance ?

Reply:
I think it is not super useful but I don't see any reason for not supporting it.

As an example, let's imagine a board with two distinct arrays of LEDs. If each array can have its own led-gpio node, then the LED indexes (or IDs) will reflect their hardware position in the array. It could be more meaningful for a user than having a big single node with the LED indexes not related with the hardware.

Reply:
In my opinion we can just support one instance to start with and make the driver multi-instance later on if we need to, since we don't have a use case for more right now.

Nothing will break and it will be totally backwards compatible as long as we access the devices using label properties as discussed elsewhere in these comments.